Ensure that subpartition template only appears after subpartition by

joaopapereira opened the pull request:
This commit changed the way the grammar was parsing the subpartition
information to ensure that the templates could not exist without a
`SUBPARTITION BY`.

Testing coverage for the case where the `SUBPARTITION TEMPLATE`
expression is written before a `SUBPARTITION BY` was added.

The error displayed when the template appeared after a Partition was
changed to point to the TEMPLATE

Add a deprecation warning to the following syntax as it should be
removed in the next major version of Greenplum.

Deprecated syntax:
 - A comma (",") between `SUBPARTITION TEMPLATE` and its
 corresponding `SUBPARTITION BY` that precedes it.
 - A comma between two consecutive `SUBPARTITION BY` clauses
---

PR for `master` branch that fixes the same issues as #5544 does to `5X_STABLE`

ashwinstar commented on src/test/regress/sql/partition.sql:
```diff
@@ -3869,3 +3869,16 @@ select gp_segment_id, attrelid::regclass, attnum, attoptions from gp_dist_random
 select oid::regclass, relkind, relstorage, reloptions from pg_class where oid = 'pt_tab_encode_1_prt_s_abc'::regclass;
 select oid::regclass, relkind, relstorage, reloptions from pg_class where oid = 'pt_tab_encode_1_prt_s_xyz'::regclass;
+
+-- MPP-26829
```
instead of number, should be adding detailed description of the issue (we can still have the number if helps for referencing).

joaopapereira commented:
We made quite a few changes to this patch set so now it's different from when it got approved. @ashwinstar or anyone else please give it another look when you have time.

User-invisible changes:
1. Further cleaned up the grammar, removed dead code

User-visible changes since last reviewed:
1. We added back compatibility for the undocumented commas
1. We added deprecation warnings

danielgustafsson commented on src/backend/parser/gram.y:
```diff
@@ -4981,57 +4980,126 @@ TabSubPartitionTemplate:
+									"corresponding SUBPARTITION TEMPLATE was "
+									"mistakenly permitted in previous versions "
+									"of Greenplum. It is deprecated and will "
+									"be removed in the next major release"),
```
AFAICT this was never documented or published in any form other than source code, do we need to throw a warning and deprecate rather than just outright not support it and error out? If the above deprecation notice stays it should be split into a short `errmsg()` with the remaining information as an `errhint()`. Something along the lines of this:
```C
ereport(WARNING,
              errmsg("comma between SUBPARTITION BY and SUBPARTITION TEMPLATE is deprecated"),
              errhint("Previous versions of Greenplum Database allowed this syntax, it will be deprecated in the next major release"),
...
```

Eulerizeit replied:
Whether we officially and properly deprecate it I don't care. 

The discussion happened offline so let me repeat it for posterity. 

We are going to pick up the partitioning code for 7.0. Given that we are probably going to really break the API at that point the desire was to maintain the functionality, (stupid as it is) until the point when we break the API for 7. 

Essentially I felt like a clean break would be better than trickling off functionality then breaking it. 

We do know that this is being used because we needed to fix it in a Jira. If we are lucky then that is the ONLY customer doing this.

danielgustafsson replied:
There is a fine line between fixing a bug that allowed for an undocumented syntax and API compatibility for an unpublished (accidental) syntax. I'm still not convinced that carrying this kludge through 6.x is worth it, but I also don't expect to have the final say on that.

danielgustafsson commented on src/backend/parser/gram.y:
```diff
@@ -4981,57 +4980,126 @@ TabSubPartitionTemplate:
+									"SUBPARTITION BY clauses was mistakenly "
+									"permitted in previous versions of "
+									"Greenplum. It is deprecated and will be "
+									"removed in the next major release"),
```
The same comment as above goes for this one, do we really need to jump through hoops to support unpublished undefined behaviour? If we do, the error message needs a bit of massaging to fit the error message standard though.

d replied:
You have my vote to gut that ...

joaopapereira commented:
After some offline talk we decided that the deprecation warning should be removed, because they were not document.

asimrp commented:
Just FYI: PR pipeline seems to indicate a legitimate failure.
